### PR TITLE
Allow partial profile information in case of exceptions

### PR DESCRIPTION
**Pull request by fjetter**
I sometimes observe timeout errors when collecting the profile information on large, busy distributed clusters. Since it's a statistical profiling I would usually not care if a fraction of the workers timeout and think this should be handled a bit more gracefully. This implementation would still return a result even in case the individual profile fetches timeout.

### CODE REVIEW

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -1234,6 +1255,42 @@ async def test_profile_metadata_keys(c, s, a, b):
...
+        "distributed.worker.profile.cycle": "100ms",
+    },
+)
+async def test_statistical_profiling(c, s, a, b):
+    futures = c.map(slowinc, range(10), delay=0.1)
+
+    await wait(futures)
+
+    profile = await s.get_profile()
+    assert profile["count"]
```
Comment (mrocklin) [inline distributed/tests/test_scheduler.py:1271]: Why this test?  Didn't we already test this functionality?

File: distributed/scheduler.py [hunk 2]
```diff
@@ -4997,9 +5000,11 @@ async def get_profile_metadata(
         else:
             workers = set(self.workers) & set(workers)
         results = await asyncio.gather(
-            *(self.rpc(w).profile_metadata(start=start, stop=stop) for w in workers)
+            *(self.rpc(w).profile_metadata(start=start, stop=stop) for w in workers),
+            return_exceptions=True
         )
 
+        results = [r for r in results if not isinstance(r, Exception)]
```
Comment (mrocklin) [inline distributed/scheduler.py:5007]: Are there ever circumstances where everything fails and we should report it?  "No" is a fine answer, I'm just bringing this up in case we can think of something.

File: distributed/tests/test_scheduler.py (hunk 1 above)
Comment (fjetter) [reply distributed/tests/test_scheduler.py:1271]: Looks like we don't have any explicit unit test for `Scheduler.get_profile` but rather a bunch of `Client.profile` tests which essentially just call into the handler.
I can understand both perspectives since the client is user facing API but otoh the dashboard calls Scheduler.get_profile directly.

File: distributed/tests/test_scheduler.py [hunk 3]
```diff
@@ -1221,6 +1221,27 @@ async def test_profile_metadata(c, s, a, b):
     assert not meta["counts"][-1][1]
 
 
+@gen_cluster(client=True, worker_kwargs={"profile_cycle_interval": 100})
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/tests/test_scheduler.py:1224]: sure

File: distributed/scheduler.py (hunk 2 above)
Comment (fjetter) [reply distributed/scheduler.py:5007]: Well, if we assume that any sort of code changes will be caught by code and we restrict ourselves to runtime failures I tend to say no (apart from systematic failure of network or similar catastrophic events). I wouldn't be surprised to be proved wrong, though. Shall we add at least some kind of logging/warning?

File: distributed/tests/test_scheduler.py [hunk 4]
```diff
@@ -1221,6 +1221,27 @@ async def test_profile_metadata(c, s, a, b):
...
+async def test_profile_metadata_timeout(c, s, a, b):
+    start = time() - 1
+
+    def raise_timeout(*args, **kwargs):
+        raise TimeoutError
+
+    b.handlers["profile_metadata"] = raise_timeout
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/tests/test_scheduler.py:1231]: Thx. This is something I actually wasn't sure about if this would be "OK". Essentially its monkeypatching which relies on knowledge of the implementation... But I figured the handler concept is pretty stable by now

**Comment by fjetter** [comment]
Since I copied the `profile_cycle_interval` I fixed the lack of explicit ms in all occurrences. Blows up the change a bit but I guess this is not harmful

**Comment by fjetter** [comment]
Test failures _seem_ to be unrelated with  `OSError: [Errno 98] Address already in use` in `test_repr_localcluster`

File: distributed/tests/test_scheduler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/tests/test_scheduler.py:1231]: Yeah, it's unfortunate in that if someone changes the implementation they now need to understand this test.  In this case though I'm not sure how else to test this functionality.  Any thoughts?

**Comment by mrocklin** [comment]
> Since I copied the profile_cycle_interval I fixed the lack of explicit ms in all occurrences. Blows up the change a bit but I guess this is not harmful

Thanks

File: distributed/tests/test_scheduler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/tests/test_scheduler.py:1231]: The only thing I can come up with which doesn't require any patching would be to continuously collect profiling information and killing one of the workers while doing so in hopes that the get_profile hits the sweet spot between "worker is dead but scheduler doesn't know about it, yet". I don't really like these timing based things unless absolutely necessary.

I assume the only implementation internal knowledge required here is:
1. There is the concept of a RPC handler
2. This handler is registered using a setitem on the attribute handler
3. The handlers name is profile_metadata

I'd assume 1 is a given and won't change as long as this project exist. Even if it did, I'd expect that these tests would be the least of the problems encountered when trying to change something that fundamental. If 2. or 3. is changed, it is at least not necessary to change or understand the test logic but only to change the assignment here.

TL;DR I believe this is a valid case of patching/mocking, just wanted to bring it up in case somebody else has an issue with this or a better idea

**Comment by mrocklin** [comment]
I apologize for letting this slip.  @fjetter feel free to merge if you think that this is ready.

**Comment by mrocklin** [comment]
Thanks Florian!
